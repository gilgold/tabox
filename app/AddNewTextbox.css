input[type="text"] {
  font-size: .85em;
  background-color: var(--input-bg-color);
  color: var(--text-color);
  padding: 4px 10px 7px 5px;
  float: left;
  width: 290px;
  border: none;
  border-bottom: 1px solid var(--primary-color);
  padding-right: 25px;
}

input[type="text"]:focus {
  outline: none;
}

label.textbox_label {
  color: var(--text-color);
  font-size: .85em;
  font-weight: normal;
  position: absolute;
  pointer-events: none;
  left: 5px;
  top: 6px;
  transition: 0.25s ease-in-out all;
}

input[type="text"]:focus~label.textbox_label, input[type="text"]:not(:placeholder-shown)~label.textbox_label {
  top: -20px;
  left: 0;
  font-size: .75em;
  color: var(--text-color);
}

.bar {
  float: left;
  position: relative;
  display: block;
  width: 290px;
}

.bar:before, .bar:after {
  content: '';
  height: 2px;
  width: 0;
  bottom: 1px;
  position: absolute;
  background: var(--primary-color);
  transition: 0.3s ease all;
  -moz-transition: 0.3s ease all;
  -webkit-transition: 0.3s ease all;
}

.bar:before {
  left: 50%;
}

.bar:after {
  right: 50%;
}

/* active state */

input[type="text"]:focus~.bar:before, input[type="text"]:focus~.bar:after {
  width: 50%;
}

/* buttons */

.btn {
  position: relative;
  margin-left: 5px;
  display: inline-block;
  padding: 0;
  overflow: hidden;
  border-width: 0;
  outline: none;
  border-radius: 2px;
  box-shadow: 0 1px 4px rgba(0, 0, 0, .6);
  background-color: var(--primary-color);
  color: #ecf0f1;
  transition: background-color .3s;
}

.btn:hover, .btn:focus {
  background-color: #21a77e;
  cursor: pointer;
}

.btn>* {
  position: relative;
}

.btn span {
  display: inline-block;
  padding: 6px 18px;
  font-size: 1em;
  font-weight: 600;
}

.btn:before {
  content: "";
  position: absolute;
  top: 50%;
  left: 50%;
  display: block;
  width: 0;
  padding-top: 0;
  border-radius: 100%;
  background-color: rgba(236, 240, 241, .3);
  transform: translate(-50%, -50%);
}

.btn:active:before {
  width: 120%;
  padding-top: 120%;
  transition: width .2s ease-out, padding-top .2s ease-out;
}

.highlighted_note {
  font-size: .65em;
  top: 3px;
  right: 7.2rem;
  color: var(--text-color);
  display: none;
  word-break: break-word;
  white-space: normal;
  width: 110px;
  vertical-align: middle;
  margin-left: 5px;
}

.highlighter {
  font-weight: bolder;
}

.clear-button {
  position: absolute;
  top: 5px;
  right: 249px;
  border: none;
  background-color: transparent;
  color: var(--text-color);
  z-index: 10;
  cursor: pointer;
  transition: opacity .2s;
}