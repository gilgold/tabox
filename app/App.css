.main-content-wrapper {
  padding: 0 5px 5px 5px;
  overflow: hidden;
  background-color: var(--bg-color);
}
.hr {
	height: 1px;
	border-top: 1px solid var(--divider-color);
	margin: 3px 0;
}

p {
	margin:0;
}