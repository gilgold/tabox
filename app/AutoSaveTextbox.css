.autosave-wrapper {
    width: 160px;
    position: relative;
    padding-left: 2px;
    display: flex;
    align-items: center;
    height: 22px;
    display: inline-flex;
}
.edit-icon {
    position: absolute;
    left: 1px;
    top: 3px;
    width: 17px;
    height: 17px;
}
.autosave-textbox {
    width: 155px !important;
    height: 20px;
    padding: 11px 20px;
    font-size: .8rem !important;
    background-color: transparent;
    border: none;
    color: var(--text-color);
    font-weight: 400;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
}
.autosave-textbox:focus, .autosave-textbox:hover {
    background-color: var(--bg-color);
    outline: none;
}

.checkmark__circle {
    stroke-dasharray: 166;
    stroke-dashoffset: 166;
    stroke-width: 2;
    stroke-miterlimit: 10;
    stroke: #7ac142;
    fill: none;
    animation: stroke 0.6s cubic-bezier(0.65, 0, 0.45, 1) forwards;
  }
  
  .checkmark {
    width: 15px;
    height: 15px;
    position: absolute;
    top: calc(50% - 7px);
    right: 5px;
    border-radius: 50%;
    display: block;
    stroke-width: 2;
    stroke: #fff;
    stroke-miterlimit: 10;
    box-shadow: inset 0px 0px 0px #7ac142;
    animation: fill .4s ease-in-out .4s forwards, scale .3s ease-in-out .9s both;
  }
  
  .checkmark__check {
    transform-origin: 50% 50%;
    stroke-dasharray: 48;
    stroke-dashoffset: 48;
    animation: stroke 0.3s cubic-bezier(0.65, 0, 0.45, 1) 0.8s forwards;
  }
  
  @keyframes stroke {
    100% {
      stroke-dashoffset: 0;
    }
  }
  @keyframes scale {
    0%, 100% {
      transform: none;
    }
    50% {
      transform: scale3d(1.1, 1.1, 1);
    }
  }
  @keyframes fill {
    100% {
      box-shadow: inset 0px 0px 0px 30px #7ac142;
    }
  }

  .lds-ring {
    display: inline-block;
    position: absolute;
    top: calc(50% - 6px);
    right: 5px;
    width: 15px;
    height: 15px;
  }
  .lds-ring div {
    box-sizing: border-box;
    display: block;
    position: absolute;
    width: 10px;
    height: 10px;
    margin: 1px;
    border: 1px solid var(--text-color);
    border-radius: 50%;
    animation: lds-ring 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
    border-color: var(--text-color) transparent transparent transparent;
  }
  .lds-ring div:nth-child(1) {
    animation-delay: -0.45s;
  }
  .lds-ring div:nth-child(2) {
    animation-delay: -0.3s;
  }
  .lds-ring div:nth-child(3) {
    animation-delay: -0.15s;
  }
  @keyframes lds-ring {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }