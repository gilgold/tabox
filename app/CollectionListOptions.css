.options-wrapper {
	display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: flex-start;
	background-color: var(--input-bg-color);
}
.sortLabel {
	color: var(--text-color);
	font-weight: 700;
	font-size: .7rem;
}

.sort-select-container {
	color: var(--settings-row-text-color);
	background-color: var(--input-bg-color);
	border: 1px solid transparent;
	display: inline-flex;
	align-items: center;
	justify-content: center;
	font-weight: 400;
	margin: 5px 5px;
	cursor: pointer;
	line-height: 16px;
	font-size: .7rem;
	transition: all .2s ease-in !important;
}

.sort-select-custom-option {
	display: flex;
	justify-content: flex-start;
}

.sort-select__control {
	color: var(--settings-row-text-color) !important;
	background-color: var(--input-bg-color) !important;
}

.sort-select__indicator svg {
	fill: var(--settings-row-text-color) !important;
}

.sort-select__option {
	background-color: var(--input-bg-color) !important;
}

.sort-select__option:hover {
	background-color: var(--single-tab-row-hover-color) !important;
}

.sort-select__menu {
	background-color: var(--input-bg-color) !important;
}

.sort-select__single-value {
	color: var(--settings-row-text-color) !important;
	min-width: 40px !important;
}

.sort-select__option--is-selected {
	color: var(--primary-color) !important;
	font-weight: 700 !important;
}

.sort-icon:before {
	content: '';
	width: 9px;
	border-top: 1px solid var(--text-color);
	position: absolute;
	top: 2px;
	height: 1px;
}

.sort-icon {
	position: relative;
	display: inline-block;
	width: 12px;
	height: 7px;
	margin: 0 5px;
	border-top: 1px solid var(--text-color);
}

.sort-icon:after {
	content: '';
	position: absolute;
	top: 5px;
	left: 0;
	width: 6px;
	height: 1px;
	border-top: 1px solid var(--text-color);
}

.verticle-divider {
	width: 1px;
	height: 26px;
	background-color: var(--divider-color);
	margin: 0 12px;
}