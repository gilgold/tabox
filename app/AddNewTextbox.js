import React, { useEffect, useRef, useState } from 'react';
import { useSetRecoilState, useRecoilValue } from 'recoil';
import './AddNewTextbox.css';
import { settingsDataState } from './atoms/settingsDataState';
import { isHighlightedState } from './atoms/globalAppSettingsState';
import { rowToHighlightState } from './atoms/animationsState';
import { getCurrentTabsAndGroups } from './utils';
import { browser } from '../static/globals';
import { useSnackbar } from 'react-simple-snackbar';
import { SnackbarStyle } from './model/SnackbarTypes';


function SaveHighlightedOnlyLabel(props) {
    const isHighlighted = useRecoilValue(isHighlightedState);
    const [totalHighlighted, setTotalHighlighted] = useState(0);

    useEffect(async () => {
        const total = (await browser.tabs.query({highlighted:true})).length;
        setTotalHighlighted(total);
    }, [])

    return <span className="highlighted_note" style={{display:(isHighlighted ? 'inline-block' : 'none')}}>Save <span className="highlighter">{totalHighlighted} selected</span> tabs</span>
}

const useFocus = () => {
    const htmlElRef = useRef(null)
    const setFocus = () => {htmlElRef.current && htmlElRef.current.focus()}

    return [ htmlElRef, setFocus ]
}

function AddNewTextbox(props) {

    const [collectionName, setName] = useState("");
    const [disabled, setDisabled] = useState(false);
    const [inputRef, setInputFocus] = useFocus();
    const settingsData = useRecoilValue(settingsDataState);
    const isHighlighted = useRecoilValue(isHighlightedState);
    const setRowToHighlight = useSetRecoilState(rowToHighlightState);
    const [openSnackbar, closeSnackbar] = useSnackbar({style: SnackbarStyle.ERROR});

    useEffect(() => {
        setInputFocus();
    }, [])


    async function handleSave() {
        if (collectionName.trim() === '') {
            openSnackbar('Please enter a name for the collection', 2000);
            return;
        }
        setDisabled(true);
        const newItem = await getCurrentTabsAndGroups(collectionName, isHighlighted);
        const newSettingsData = settingsData ? [newItem, ...settingsData] : [newItem];
        setRowToHighlight(0);
        await props.updateRemoteData(newSettingsData);
        setTimeout(() => setDisabled(false), 1000);
    }

    async function _handleKeyDown(e) {
        if (e.key === 'Enter') {
            await handleSave();
        }
      }

    return <section>
                <div className="group">
                    <input
                        type="text"
                        maxLength="50"
                        placeholder=" "
                        name="new_setting_title"
                        id="new_setting_title"
                        onKeyDown={async e => await _handleKeyDown(e)}
                        onChange={e => setName(e.target.value)}
                        ref={inputRef}
                        style={{
                            float: 'left',
                            width: '290px',
                        }}
                    />
                    <span
                        className="bar"
                        style={{
                            float: 'left',
                            width: '290px',
                        }}></span>
                    <label className="textbox_label">Save {isHighlighted ? 'selected' : 'all'} tabs as...</label>
                </div>
                <button
                    id="add_new_setting"
                    disabled={disabled}
                    className="btn"
                    onClick={async () => await handleSave()}
                    >
                        <span>Add</span>
                    </button>
                <SaveHighlightedOnlyLabel />
            </section>;
};

export default AddNewTextbox;