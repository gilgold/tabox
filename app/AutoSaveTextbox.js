import React, { useState, useEffect, useRef } from 'react';
import './AutoSaveTextbox.css';

const TypingLoader = () => <div className="lds-ring"><div /><div /><div /><div /></div>;

export const AutoSaveTextbox = (props) => {
    const [value, setValue] = useState(props.initValue);
    const [saved, setSaved] = useState(false);
    const [typing, setTyping] = useState(false);
    const [isInitial, setIsInitial] = useState(true);
    const inputRef = useRef();

    const handleOnChange = (event) => {
        setSaved(false);
        if (props.onChange) props.onChange(event.target.value);
        setValue(event.target.value);
    };

    useEffect(() => {
        setValue(props.initValue);
        setIsInitial(false);
    }, []);

    useEffect(() => {
        setIsInitial(true);
        setValue(props.initValue);
        const isInitialTimer = setTimeout(() => setIsInitial(false), 200);
        return () => {
            clearTimeout(isInitialTimer);
            setTyping(false);
            setSaved(false);
        };
    }, [props.initValue]);

    useEffect(() => {
        if (isInitial) return;
        setTyping(true);
        let typingTimer;
        let savedTimer;    
        const timeoutId = setTimeout(() => {
            props.action(value.trim(), props.item);
            typingTimer = setTimeout(() => {setTyping(false); setSaved(true);}, 100);
            savedTimer = setTimeout(() => setSaved(false), 2300);
        }, 700);
        return () => {
            clearTimeout(typingTimer);
            clearTimeout(savedTimer);
            clearTimeout(timeoutId);
        };
    }, [value]);

    return (
        <div className="autosave-wrapper">
            <div className="edit-icon" onClick={() => inputRef.current.focus()}>
                <svg width="17" height="17" viewBox="0 0 17 17" fill="none" xmlns="http://www.w3.org/2000/svg">
                <path d="M2.125 12.2187V14.8749H4.78125L12.6154 7.04077L9.95917 4.38452L2.125 12.2187ZM14.6696 4.98661C14.9458 4.71036 14.9458 4.26411 14.6696 3.98786L13.0121 2.33036C12.7358 2.05411 12.2896 2.05411 12.0133 2.33036L10.7171 3.62661L13.3733 6.28286L14.6696 4.98661V4.98661Z" fill="var(--text-color)"/>
                </svg>
            </div>
            <input 
                ref={inputRef} 
                className="autosave-textbox" 
                placeholder="No name" 
                maxLength={props.maxLength ?? -1}
                onChange={handleOnChange} 
                data-multiline={true} 
                data-tip="Type here to edit.<br />Changes are saved automatically." 
                data-class="small-tooltip"
                value={value} />
            {saved ? (
                <svg key={`saved-${saved}`} className="checkmark" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 52 52">
                    <circle className="checkmark__circle" cx="26" cy="26" r="25" fill="none" />
                    <path className="checkmark__check" fill="none" d="M14.1 27.2l7.1 7.2 16.7-16.8" />
                </svg>
            ) : typing ? <TypingLoader /> : null }
        </div>
    );
};