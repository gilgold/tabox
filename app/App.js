/* eslint-disable no-useless-escape */
import './App.css';
import React, { useEffect, useMemo, useState } from 'react';
import Header from './Header';
import AddNewTextbox from './AddNewTextbox';
import CollectionList from './CollectionList';
import Footer from './Footer';
import { useRecoilState, useSetRecoilState, useRecoilValue } from 'recoil';
import { settingsDataState } from './atoms/globalAppSettingsState';
import { applyUid } from './utils';
import {
  themeState,
  isLoggedInState,
  syncInProgressState,
  lastSyncTimeState,
  searchState,
  listKeyState,
} from './atoms/globalAppSettingsState';
import { rowToHighlightState } from './atoms/animationsState';
import { browser } from '../static/globals';
import { useSnackbar } from 'react-simple-snackbar';
import { SnackbarStyle } from './model/SnackbarTypes';
import ReopenLastSession from './ReopenLastSession';
import ReactTooltip from 'react-tooltip';
import { CollectionListOptions } from './CollectionListOptions';

const Divder = () => <div className='hr' />;

function App() {
  const [settingsData, setSettingsData] = useRecoilState(settingsDataState);
  const setRowToHighlight = useSetRecoilState(rowToHighlightState);
  const [themeMode, setThemeMode] = useRecoilState(themeState);
  const [isLoggedIn, setIsLoggedIn] = useRecoilState(isLoggedInState);
  const setSyncInProgress = useSetRecoilState(syncInProgressState);
  const setLastSyncTime = useSetRecoilState(lastSyncTimeState);
  const [openSnackbar] = useSnackbar({ style: SnackbarStyle.ERROR });
  const search = useRecoilValue(searchState);
  const [listKey, setListKey] = useRecoilState(listKeyState);
  const [sortValue, setSortValue] = useState(null);

  const removeInactiveWindowsFromAutoUpdate = async () => {
    let { collectionsToTrack } = await browser.storage.local.get('collectionsToTrack');
    const { chkEnableAutoUpdate } = await browser.storage.local.get('chkEnableAutoUpdate');
    if (!collectionsToTrack || collectionsToTrack.length === 0 || !chkEnableAutoUpdate) { return; }
    const activeWindowIds = (await browser.windows.getAll({ populate: false })).map(c => c.id);
    collectionsToTrack = collectionsToTrack.filter(c => activeWindowIds.includes(c.windowId));
    console.log('collectionsToTrack', collectionsToTrack);
    await browser.storage.local.set({ collectionsToTrack: collectionsToTrack });
  }

  const applyTheme = async () => {
    let { theme } = await browser.storage.local.get('theme');
    theme = theme ? theme : window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
    setThemeMode(theme);
    document.documentElement.setAttribute('data-theme', theme);
  }

  const checkSyncStatus = async () => {
    console.log('check sync status')
    const { googleRefreshToken } = await browser.storage.local.get('googleRefreshToken');
    if (!googleRefreshToken) return;
    browser.runtime.sendMessage({ type: 'checkSyncStatus' }).then(async (response) => {
      setIsLoggedIn(response === null ? false : response);
      if (response) await applyDataFromServer();
    });
  }

  const _handleSyncError = async () => {
    await browser.storage.local.remove('googleToken');
    await browser.storage.local.remove('googleUser');
    setIsLoggedIn(false);
    openSnackbar('Error syncing data, please enable sync again', 6000);
  }

  const logout = async () => {
    browser.runtime.sendMessage({ type: 'logout' }).then(() => {
      setIsLoggedIn(false);
    })
  };

  const applyDataFromServer = async (force = false) => {
    setSyncInProgress(true);
    browser.runtime.sendMessage({ type: 'loadFromServer', force: force }).then((response) => {
      if (response !== false) {
        setSettingsData(response);
        setLastSyncTime(Date.now());
      }
    }).catch(async (err) => {
      await _handleSyncError(err)
    }).finally(() => {
      setSyncInProgress(false);
    });
  }

  const _update = async () => {
    setSyncInProgress(true);
    console.log('Update remote data');
    browser.runtime.sendMessage({ type: 'updateRemote' }).then(() => {
      setLastSyncTime(Date.now());
    }).catch(async (err) => {
      await _handleSyncError(err)
    }).finally(() => {
      setSyncInProgress(false);
    });
  }

  const updateRemoteData = async (newData) => {
    setSettingsData(newData);
    await browser.storage.local.set({ localTimestamp: Date.now() });
    await browser.runtime.sendMessage({ type: 'addCollection' });
    if (!isLoggedIn) return;
    _update();
  }

  const updateCollection = async (newCollection) => {
    const newList = [...settingsData];
    const index = newList.findIndex(c => c.uid === newCollection.uid);
    newList[index] = newCollection;
    await updateRemoteData(newList);
  }

  const removeCollection = (collectionUid) => {
    return [...settingsData].filter(c => c.uid !== collectionUid);
  }

  const addCollection = async (newCollection) => {
    const newList = settingsData ? [newCollection, ...settingsData] : [newCollection];
    setRowToHighlight(0);
    await updateRemoteData(newList);
    const { chkAutoUpdateOnNewCollection } = await browser.storage.local.get('chkAutoUpdateOnNewCollection');
    if (!chkAutoUpdateOnNewCollection) return;
    setTimeout(async () => {
      let { collectionsToTrack } = (await browser.storage.local.get('collectionsToTrack')) || [];
      const window = await browser.windows.getLastFocused({ windowTypes: ['normal'] });
      const index = collectionsToTrack.findIndex(c => c.collectionUid === newCollection.uid);
      if (index !== undefined && index > -1) {
          collectionsToTrack[index].windowId = window.id;
      } else {
          collectionsToTrack.push({
              collectionUid: newCollection.uid,
              windowId: window.id
          });
      }
      await browser.storage.local.set({ collectionsToTrack });
      setListKey(Math.random().toString(36));
    }, 1000);
  }

  const loadCollectionsFromStorage = async () => {
    const { tabsArray } = await browser.storage.local.get('tabsArray');
    let newCollections = [];
    if (tabsArray && tabsArray.length > 0) {
      tabsArray.forEach((collection) => {
        if (collection.tabs && collection.tabs.length > 0 && !('uid' in collection.tabs[0])) {
          const taboxItem = applyUid(collection);
          newCollections.push(taboxItem);
        } else {
          newCollections.push(collection);
        }
      });
    }
    setSettingsData(newCollections);
  }

  const getSelectedSort = async () => {
    const { currentSortValue } = await browser.storage.local.get('currentSortValue');
    setSortValue(currentSortValue);
  }

  useEffect(() => {
    if (isLoggedIn) loadCollectionsFromStorage();
  }, [isLoggedIn]);

  useEffect(async () => {
    await applyTheme();
    await getSelectedSort();
    await removeInactiveWindowsFromAutoUpdate();
    await checkSyncStatus();
    await loadCollectionsFromStorage();
  }, []);

  const escapeRegex = string => {
    return string.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
  }

  const collectionsToShow = useMemo(() => {
    if (!search || !search.trim() || !settingsData) return settingsData;
    const searchRegex = new RegExp(escapeRegex(search), 'i');
    return settingsData.filter(collection => collection.name.match(searchRegex));
  }, [
    search,
    settingsData,
  ]);

  return <div className="App">
    <ReactTooltip
      event={'mouseover'}
      eventOff={'click mouseout'}
      delayShow={200}
      type={themeMode === 'light' ? 'dark' : 'light'} />
    <Header
      applyDataFromServer={applyDataFromServer}
      updateRemoteData={updateRemoteData}
      logout={logout} />
    <div className="main-content-wrapper">
      <AddNewTextbox addCollection={addCollection} updateRemoteData={updateRemoteData} />
      <Divder />
      <CollectionListOptions 
        key={`${sortValue}-select`}
        updateRemoteData={updateRemoteData} 
        selected={sortValue}
      />
      <Divder />
      <CollectionList
        key={`collection-list-${listKey}`}
        updateRemoteData={updateRemoteData}
        collections={collectionsToShow}
        updateCollection={updateCollection}
        removeCollection={removeCollection} />
      <Divder />
      <ReopenLastSession addCollection={addCollection} />
    </div>
    <Footer />
  </div>;
}

export default App;
