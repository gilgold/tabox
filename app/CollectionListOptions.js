import React from 'react';
import { useRecoilValue } from 'recoil';
import { settingsDataState } from './atoms/globalAppSettingsState';
import './CollectionListOptions.css';
import Switch from './Switch';
import { IoColorPalette } from 'react-icons/io5';
import { FaCalendarAlt } from 'react-icons/fa';


export function CollectionListOptions(props) {
    const settingsData = useRecoilValue(settingsDataState);

    const SortType = {
        COLOR: (a, b) => (b.color > a.color) ? 1 : ((a.color > b.color) ? -1 : 0),
        DATE: (a, b) => b.createdOn - a.createdOn
    }

    const handleSort = async (sortBy) => {
        if (!settingsData || settingsData.length === 0) return;
        let newSettingsData = [...settingsData];
        newSettingsData.sort(sortBy)
        await props.updateRemoteData(newSettingsData);
    }

    return <div className="options-wrapper">
        <div className="sort-icon" /> <span className="sortLabel">Sort by: </span>
        <button className="button" onClick={async () => handleSort(SortType.COLOR)}><IoColorPalette />&nbsp; Color</button>
        <button className="button" onClick={async () => handleSort(SortType.DATE)}><FaCalendarAlt />&nbsp; Newest</button>
        <div className="verticle-divider" />
        <Switch 
            id="chkOpenNewWindow" 
            textOn={<span>Tabs will open in a <strong>new window</strong></span>} 
            textOff={<span>Tabs will be added to <strong>this window</strong></span>} />
    </div>;
}
