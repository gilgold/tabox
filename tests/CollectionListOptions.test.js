import React from 'react';
import renderer from 'react-test-renderer';
import { CollectionListOptions } from '../app/CollectionListOptions';
import { RecoilRoot } from 'recoil';

describe('Collection List Options tests', () => {
  test('CollectionListOptions renders correctly', () => {
    const component = renderer.create(
      <RecoilRoot>
        <CollectionListOptions />
      </RecoilRoot>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});