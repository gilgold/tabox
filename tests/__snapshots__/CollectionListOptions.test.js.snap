// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection List Options tests CollectionListOptions renders correctly 1`] = `
<div
  className="options-wrapper"
>
  <div
    className="sort-icon"
  />
   
  <span
    className="sortLabel"
  >
    Sort by: 
  </span>
  <button
    className="button"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M441 336.2l-.06-.05c-9.93-9.18-22.78-11.34-32.16-12.92l-.69-.12c-9.05-1.49-10.48-2.5-14.58-6.17-2.44-2.17-5.35-5.65-5.35-9.94s2.91-7.77 5.34-9.94l30.28-26.87c25.92-22.91 40.2-53.66 40.2-86.59s-14.25-63.68-40.2-86.6c-35.89-31.59-85-49-138.37-49C223.72 48 162 71.37 116 112.11c-43.87 38.77-68 90.71-68 146.24s24.16 107.47 68 146.23c21.75 19.24 47.49 34.18 76.52 44.42a266.17 266.17 0 0086.87 15h1.81c61 0 119.09-20.57 159.39-56.4 9.7-8.56 15.15-20.83 15.34-34.56.21-14.17-5.37-27.95-14.93-36.84zM112 208a32 32 0 1132 32 32 32 0 01-32-32zm40 135a32 32 0 1132-32 32 32 0 01-32 32zm40-199a32 32 0 1132 32 32 32 0 01-32-32zm64 271a48 48 0 1148-48 48 48 0 01-48 48zm72-239a32 32 0 1132-32 32 32 0 01-32 32z"
      />
    </svg>
      Color
  </button>
  <button
    className="button"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V192H0v272zm320-196c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM192 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM64 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zM400 64h-48V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H160V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H48C21.5 64 0 85.5 0 112v48h448v-48c0-26.5-21.5-48-48-48z"
      />
    </svg>
      Newest
  </button>
  <div
    className="verticle-divider"
  />
  <span
    data-class="small-tooltip"
    data-multiline={true}
  >
    <input
      checked={false}
      className="switch-input"
      id="chkOpenNewWindow"
      name="chkOpenNewWindow"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="switch-label"
      htmlFor="chkOpenNewWindow"
    >
      <span
        className="toggle--on"
      >
        <span>
          Tabs will open in a 
          <strong>
            new window
          </strong>
        </span>
      </span>
      <span
        className="toggle--off"
      >
        <span>
          Tabs will be added to 
          <strong>
            this window
          </strong>
        </span>
      </span>
    </label>
  </span>
</div>
`;
